inputs/loadgenerator/core.go

```diff
@@ -8,58 +8,253 @@ import (
 	"context"
 	"fmt"
 	"os"
+	"sync"
 	"time"
 
+	"github.com/elastic/elastic-agent-client/v7/pkg/client"
+	"github.com/elastic/elastic-agent-libs/config"
 	"github.com/elastic/elastic-agent-libs/logp"
+
+	"go.uber.org/atomic"
+	"google.golang.org/protobuf/types/known/structpb"
 )
 
 type loadGenerator struct {
-	cfg    Config
-	now    func() time.Time
-	logger *logp.Logger
-}
-
-func (l loadGenerator) Run(ctx context.Context) error {
-	t := l.now()
-	delta := l.cfg.TimeDelta
-	delay := l.cfg.Delay
-
-	// Loop forever
-	if l.cfg.Loop {
-		l.logger.Debug("loop enabled")
-		for {
-			select {
-			case <-ctx.Done():
-				l.logger.Info("Shutdown signal received, closing loadGenerator")
-				return nil
-			default:
-				if err := l.send(l.next(t)); err != nil {
-					return fmt.Errorf("error sending event: %w", err)
+	now     func() time.Time
+	logger  *logp.Logger
+	runners map[string]*loadRunner
+	runMut  *sync.Mutex
+	events  atomic.Uint64
+}
+
+func newLoadGenerator(logger *logp.Logger) *loadGenerator {
+
+	return &loadGenerator{
+		now:     time.Now,
+		logger:  logger,
+		runners: make(map[string]*loadRunner),
+		runMut:  &sync.Mutex{},
+	}
+}
+
+// load runner handles the communications between a loadGenerator instance and the server
+type loadRunner struct {
+	stopFuncs map[string]chan struct{}
+	mapMut    *sync.Mutex
+	unit      *client.Unit
+}
+
+func newLoadRunner() *loadRunner {
+	return &loadRunner{
+		stopFuncs: make(map[string]chan struct{}),
+		mapMut:    &sync.Mutex{},
+	}
+}
+
+func (l *loadGenerator) Close(ctx context.Context) error {
+	// nothing here, for now
+	return nil
+}
+
+// Start the generator and initialize the client
+func (l *loadGenerator) Start(ctx context.Context) error {
+	agentClient, _, err := client.NewV2FromReader(os.Stdin, client.VersionInfo{
+		Name:    "beat-v2-client",
+		Version: "alpha",
+		Meta:    map[string]string{},
+	})
+	if err != nil {
+		return fmt.Errorf("error fetching client from stdin: %w", err)
+	}
+	return l.StartWithClient(ctx, agentClient)
+}
+
+// StartWithClient initializes the client and starts the main runloop of the generator
+func (l *loadGenerator) StartWithClient(ctx context.Context, agentClient client.V2) error {
+	err := agentClient.Start(ctx)
+	if err != nil {
+		return fmt.Errorf("error starting client: %w", err)
+	}
```

Comment:
I think that's just a personal preference, and we usually don't combine the two in the beats codebase, at least in my experience. Generally not a fan of `if` statements with side effects, since it's a little bit easier for my brain to skip over.